src/frontend/src/lib/utils/token.utils.ts

```diff
@@ -131,3 +141,100 @@ export const mapTokenUi = ({
+ * This checks whether the token has a twinTokenSymbol field and ensures that it is a string.
+ *
+ * @param token - The token object to be checked.
+ * @returns A boolean indicating whether the token is of type RequiredTokenWithLinkedData.
+ */
+export function isRequiredTokenWithLinkedData(token: Token): token is RequiredTokenWithLinkedData {
+	return 'twinTokenSymbol' in token && typeof token.twinTokenSymbol === 'string';
+}
+
+/**
+ * Type guard to check if an object is of type TokenGroupUi.
+ *
+ * @param obj - The object to check.
+ * @returns A boolean indicating whether the object is a TokenGroupUi.
+ */
+export function isTokenGroupUi(obj: unknown): obj is TokenGroupUi {
+	return (
+		typeof obj === 'object' &&
+		obj !== null &&
+		'header' in obj &&
+		typeof (obj as TokenGroupUi).header === 'object' &&
+		'tokens' in obj
+	);
+}
+
+/**
+ * Factory function to create a TokenGroupUi based on the provided tokens and network details.
+ * This function creates a group header and adds both the native token and the twin token to the group's tokens array.
+ *
+ * @param nativeToken - The native token used for the group, typically the original token or the one from the selected network.
+ * @param twinToken - The twin token to be grouped with the native token, usually representing the same asset on a different network.
+ *
+ * @returns A TokenGroupUi object that includes a header with network and symbol information and contains both the native and twin tokens.
+ */
+function createTokenGroup({
+	nativeToken,
+	twinToken
+}: {
+	nativeToken: TokenUi;
+	twinToken: TokenUi;
+}): TokenGroupUi {
+	const capitalizeNetworkName = (name: string) => name.charAt(0).toUpperCase() + name.slice(1);
+
+	return {
+		header: {
+			name: capitalizeNetworkName(nativeToken.network.name),
+			symbol: `${nativeToken.symbol}, ${twinToken.symbol}`,
+			decimals: nativeToken.decimals,
+			icon: nativeToken.icon ?? '/images/default_token_icon.svg'
+		},
+		native: nativeToken.network,
+		tokens: [nativeToken, twinToken]
+	};
+}
+
+/**
+ * Function to create a list of TokenUiOrGroupUi by grouping tokens with matching twinTokenSymbol.
+ * The group is placed in the position where the first token of the group was found.
+ * Tokens with no twin remain as individual tokens in their original position.
+ *
+ * @param tokens - The list of TokenUi objects to group. Each token may or may not have a twinTokenSymbol.
+ *                 Tokens with a twinTokenSymbol are grouped together.
+ *
+ * @returns A new list where tokens with twinTokenSymbols are grouped into a TokenGroupUi,
+ *          and tokens without twins remain in their original place.
+ *          The group replaces the first token of the group in the list.
+ */
+export function groupTokensByTwin(tokens: TokenUi[]): TokenUiOrGroupUi[] {
+	const groupedTokenTwins = new Set<string>();
+	const mappedTokensWithGroups: TokenUiOrGroupUi[] = tokens.map((token) => {
```

Comment:
may you please make it functional, maybe using reduce to define both variables `groupedTokenTwins` and `mappedTokensWithGroups`

Reply:
7bcf93a7